index.rst

```diff
@@ -4,9 +4,149 @@ File Formats and Layouts for Cell-based Coadds
+We do not expect to need the ability to extend existing files.
+
+We expect to compress all image planes with at least lossless compression, and would like to have the capability to perform lossy compression on some planes as well.
+If we do lossy compression, we would quite likely want to do our own quantization (our `afw` library already has code for this that we prefer over CFITSIO's).
+DESC has expressed an interest in applying lossy compression to any coadd files they transfer to NERSC, if we have not done so already.
+
+Constraints from FITS
+=====================
+
+Because we need to be able to read subimages efficiently, file-level compression is not an option: our only options are FITS image "tile compression" and perhaps lesser-known FITS 4.0's binary table compression.
+Combined with our need to subimage reads over object stores, this already puts us beyond what third-party FITS libraries can do:
+
+- CFITSIO can do image tile compression, including efficient subimage reads that take advantage of the tile structure, but only on POSIX filesystems (or blocks of contiguous memory that correspond to the on-disk layout).
+  It can also do subset reads of FITS binary tables, again only on POSIX filesystems.
+  This imposes the same limitation on its Python bindings in the ``fitsio`` package.
+
+- Astropy can do image tile compression with efficient subimage reads on POSIX filesystems, and it can do uncompressed reads (including efficient subimage reads) against both POSIX and object stores via `fsspec`, but it delegates to vendored CFITSIO code for its tile compression code and does not support the combination of these two.
+  Astropy cannot do subset reads of FITS binary tables, even on POSIX filesystems.
+
+- I am not aware of any library that can do FITS binary table compression at all.
+
+As a result, I expect us to need to write some low-level code of our own to do subimage reads over object stores, though this is not specific to cell-based coadds: we need this for all of our image data products.
+Whether to contribute more general code upstream (probably to Astropy) or focus only on reading the specific files we write ourselves is an important open question; the general solution would be very useful the community, but its scope is unquestionably larger.
+
+Finally, the WCS of any single cell (or the inner-cell stitched image) will be simple and can be exactly represented in FITS headers (unlike the WCSs of our single-epoch images).
+An additional FITS WCS could be used to describe the position of a single-cell image within a larger grid.
+It is highly desirable that any images we store in our FITS files have both to allow third-party FITS readers to fully interpret them.
+
+Image Data Layouts
+==================
+
+This section will cover potential ways to lay out the image data - the 5+ planes that share a common grid, as well as the PSF images - across multiple FITS HDUs.
+
+Binary Table With Per-Cell Rows
```

Comment:
I think you are aware, but the FITS tile compression code fpack ends up with a binary table of compressed images. In DES, we get pretty good (lossy) compression with the tiling working per cell. We found pretty apparent issues with lossy compression when the tile compression boundaries crossed cell boundaries because the noise levels in the images change.

Reply:
It really makes me wonder why we aren't using HDF5 and making use of a hierarchy and native compression rather than trying to shoe horn this into FITS.

Reply:
The way I see it, we can _also_ do HDF5, but we really can't _just_ do HDF5, according to our requirements.  Maybe there's some scheme where we only store the HDF5 and convert to FITS when streaming bytes in various places, but I'm not at all convinced that removes enough constraints from the FITS version to make it easier to implement overall.

Reply:
Forcing cell-based coadds into FITS when the thing you end up with is essentially unusable for ds9/firefly/astropy seems like we are stretching the usefulness of the requirement we have to use FITS. If we have a service to convert the cell-based thing to a usable image then that service returning FITS is good enough. cc/ @gpdf

Reply:
**To be very clear, the cells with their buffers are the usable images.** Any algorithm one runs on them will typically be run on the full cells with buffers and then info from the buffers is discarded at the end.

Reply:
I think there are two cases:

- if you want to do something convolution-like, you want the full cells;
- if you want to do forward modeling (or naive things that don't care about the PSF), you don't care about the outer regions.  Instead you just render a piecewise model with the piecewise PSF and compare that to the piecewise data.

While both are important, I actually think the latter will be much more common.

Reply:
Matt and Tim mean different things by the term 'usable'. Matt is referring to specific algorithms being able to run as usable, whereas Tim is referring to third-party tools able to interpret, stitch and visualize as 'usable'

> If we have a service to convert the cell-based thing to a usable image then that service returning FITS is good enough.

Interesting. Do we have a service to do a bulk data transfer? Such a conversion _could_ happen in the process, since we don't want to store the trimmed/stitched image at USDF and the destination may not have Science Pipelines installed to read our custom datastructure.

Reply:
The thing with item 2 above is that as soon as you think about how to parallelize a forward modeling algorithm, you'll end up back in the case of using some sort of buffer. The overlaps are meant to enable these things and not an annoyance that we have to work around.